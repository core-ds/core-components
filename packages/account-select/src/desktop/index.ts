export { AccountSelectDesktop } from './Component.desktop';
