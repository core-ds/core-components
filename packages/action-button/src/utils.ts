import { getDataTestId } from '@alfalab/core-components-shared';

export function getActionButtonTestIds(dataTestId: string) {
    return {
        actionButton: dataTestId,
        spinner: getDataTestId(dataTestId, 'loader'),
    };
}
