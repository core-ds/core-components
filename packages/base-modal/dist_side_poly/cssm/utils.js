'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var coreComponentsGlobalStore = require('@alfalab/core-components-global-store/cssm');
var coreComponentsShared = require('@alfalab/core-components-shared/cssm');

function isScrolledToTop(target) {
    return target.scrollTop <= 0;
}
function isScrolledToBottom(target) {
    return target.scrollHeight - target.offsetHeight <= target.scrollTop;
}
function hasScrollbar(target) {
    return target.scrollHeight > target.clientHeight;
}
var isOverflowing = function (container) {
    if (document.body === container) {
        return window.innerWidth > document.documentElement.clientWidth;
    }
    return container.scrollHeight > container.clientHeight;
};
var getPaddingRight = function (node) {
    return parseInt(window.getComputedStyle(node).paddingRight, 10) || 0;
};
var restoreContainerStyles = function (container) {
    var modalRestoreStyles = coreComponentsGlobalStore.getModalStore().getRestoreStyles();
    var index = modalRestoreStyles.findIndex(function (s) { return s.container === container; });
    var existingStyles = modalRestoreStyles[index];
    if (!existingStyles)
        return;
    existingStyles.modals -= 1;
    if (existingStyles.modals <= 0) {
        modalRestoreStyles.splice(index, 1);
        existingStyles.styles.forEach(function (_a) {
            var value = _a.value, el = _a.el, key = _a.key;
            if (value) {
                el.style.setProperty(key, value);
            }
            else {
                el.style.removeProperty(key);
            }
        });
    }
};
var handleContainer = function (container, shouldIOSLock) {
    if (shouldIOSLock === void 0) { shouldIOSLock = false; }
    if (!container)
        return;
    var modalRestoreStyles = coreComponentsGlobalStore.getModalStore().getRestoreStyles();
    var existingStyles = modalRestoreStyles.find(function (s) { return s.container === container; });
    if (existingStyles) {
        existingStyles.modals += 1;
        return;
    }
    var containerStyles = [];
    if (isOverflowing(container)) {
        // Вычисляет размер до применения `overflow hidden` для избежания скачков
        var scrollbarSize = coreComponentsShared.browser.getScrollbarSize();
        containerStyles.push({
            value: container.style.paddingRight,
            key: 'padding-right',
            el: container,
        });
        // Вычисляем стили, чтобы получить реальный `padding` c шириной сколлбара
        // eslint-disable-next-line no-param-reassign
        container.style.paddingRight = "".concat(getPaddingRight(container) + scrollbarSize, "px");
    }
    var parent = container.parentElement;
    var scrollContainer = 
    // TODO: заменить на optional chaining
    parent &&
        parent.nodeName === 'HTML' &&
        window.getComputedStyle(parent).overflowY === 'scroll'
        ? parent
        : container;
    // Блокируем скролл даже если отсутствует скроллбар
    if (scrollContainer.style.overflow !== 'hidden') {
        containerStyles.push({
            value: scrollContainer.style.overflow,
            key: 'overflow',
            el: scrollContainer,
        });
    }
    if (!shouldIOSLock) {
        scrollContainer.style.overflow = 'hidden';
    }
    modalRestoreStyles.push({
        container: container,
        modals: 1,
        styles: containerStyles,
    });
};

exports.handleContainer = handleContainer;
exports.hasScrollbar = hasScrollbar;
exports.isScrolledToBottom = isScrolledToBottom;
exports.isScrolledToTop = isScrolledToTop;
exports.restoreContainerStyles = restoreContainerStyles;
