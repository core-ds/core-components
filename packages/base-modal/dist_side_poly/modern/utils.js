import { getModalStore } from '@alfalab/core-components-global-store/modern';
import { browser } from '@alfalab/core-components-shared/modern';

function isScrolledToTop(target) {
    return target.scrollTop <= 0;
}
function isScrolledToBottom(target) {
    return target.scrollHeight - target.offsetHeight <= target.scrollTop;
}
function hasScrollbar(target) {
    return target.scrollHeight > target.clientHeight;
}
const isOverflowing = (container) => {
    if (document.body === container) {
        return window.innerWidth > document.documentElement.clientWidth;
    }
    return container.scrollHeight > container.clientHeight;
};
const getPaddingRight = (node) => parseInt(window.getComputedStyle(node).paddingRight, 10) || 0;
const restoreContainerStyles = (container) => {
    const modalRestoreStyles = getModalStore().getRestoreStyles();
    const index = modalRestoreStyles.findIndex((s) => s.container === container);
    const existingStyles = modalRestoreStyles[index];
    if (!existingStyles)
        return;
    existingStyles.modals -= 1;
    if (existingStyles.modals <= 0) {
        modalRestoreStyles.splice(index, 1);
        existingStyles.styles.forEach(({ value, el, key }) => {
            if (value) {
                el.style.setProperty(key, value);
            }
            else {
                el.style.removeProperty(key);
            }
        });
    }
};
const handleContainer = (container, shouldIOSLock = false) => {
    if (!container)
        return;
    const modalRestoreStyles = getModalStore().getRestoreStyles();
    const existingStyles = modalRestoreStyles.find((s) => s.container === container);
    if (existingStyles) {
        existingStyles.modals += 1;
        return;
    }
    const containerStyles = [];
    if (isOverflowing(container)) {
        // Вычисляет размер до применения `overflow hidden` для избежания скачков
        const scrollbarSize = browser.getScrollbarSize();
        containerStyles.push({
            value: container.style.paddingRight,
            key: 'padding-right',
            el: container,
        });
        // Вычисляем стили, чтобы получить реальный `padding` c шириной сколлбара
        // eslint-disable-next-line no-param-reassign
        container.style.paddingRight = `${getPaddingRight(container) + scrollbarSize}px`;
    }
    const parent = container.parentElement;
    const scrollContainer = 
    // TODO: заменить на optional chaining
    parent &&
        parent.nodeName === 'HTML' &&
        window.getComputedStyle(parent).overflowY === 'scroll'
        ? parent
        : container;
    // Блокируем скролл даже если отсутствует скроллбар
    if (scrollContainer.style.overflow !== 'hidden') {
        containerStyles.push({
            value: scrollContainer.style.overflow,
            key: 'overflow',
            el: scrollContainer,
        });
    }
    if (!shouldIOSLock) {
        scrollContainer.style.overflow = 'hidden';
    }
    modalRestoreStyles.push({
        container,
        modals: 1,
        styles: containerStyles,
    });
};

export { handleContainer, hasScrollbar, isScrolledToBottom, isScrolledToTop, restoreContainerStyles };
