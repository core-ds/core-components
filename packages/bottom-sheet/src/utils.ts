export const TIMEOUT = 500;
export const SWIPE_VELOCITY = 0.4;
export const MARKER_HEIGHT = 24;
/* Верхний отступ шторки, если она открыта на максимальную высоту */
export const HEADER_OFFSET = 24;
export const CLOSE_OFFSET = 0.2;
export const SCROLL_OFFSET = 1;

export const convertPercentToNumber = (value: string | number, fullHeight: number) => {
    const maxHeight = fullHeight - HEADER_OFFSET;

    if (typeof value === 'string') {
        const percent = parseFloat(value) / 100;

        if (percent < 0) {
            return Math.min(maxHeight, fullHeight + fullHeight * percent);
        }

        return Math.min(maxHeight, fullHeight * percent);
    }

    if (value < 0) {
        return Math.max(0, fullHeight + value);
    }

    return Math.min(maxHeight, value);
};
