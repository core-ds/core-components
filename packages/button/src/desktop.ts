export { ButtonDesktop } from './Component.desktop';
export type { CommonButtonProps as ButtonDesktopProps } from './typings';
