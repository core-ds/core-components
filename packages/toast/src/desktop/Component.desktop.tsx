import React, { forwardRef } from 'react';

import { ToastPlateDesktop } from '@alfalab/core-components-toast-plate/desktop';

import { BaseToast, BaseToastProps } from '../components/base-toast';

export type ToastDesktopProps = Omit<BaseToastProps, 'bottomButtonPosition' | 'breakpoint'>;

const DefaultToastPlateDesktop: BaseToastProps['ToastPlate'] = forwardRef((props, ref) => (
    <ToastPlateDesktop ref={ref} {...props} />
));

export const ToastDesktop = forwardRef<HTMLDivElement, ToastDesktopProps>(
    ({ ToastPlate = DefaultToastPlateDesktop, ...restProps }, ref) => (
        <BaseToast ToastPlate={ToastPlate} {...restProps} ref={ref} />
    ),
);
