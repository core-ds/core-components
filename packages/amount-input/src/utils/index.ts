/**
 * https://github.com/core-ds/utils
 * TODO: нужно перенести утилиты в этот проект, когда он будет готов
 */

import { CurrencyCodes } from '@alfalab/data';
import { formatAmount } from '@alfalab/utils';

/**
 * Форматирует введенное значение
 * @param enteredValue Значение введенное в инпут
 * @param currency валюта
 * @param minority количество минорных единиц
 */
export function getFormattedValue(enteredValue: string, currency: CurrencyCodes, minority: number) {
    if (!enteredValue || enteredValue === '-') {
        return enteredValue;
    }

    // eslint-disable-next-line prefer-const
    let [head, tail] = enteredValue.split(',');

    // При вводе "-," указываем, что имеется в виду "-0,"
    if (head === '-') {
        head = '-0';
    }

    let { majorPart } = formatAmount({
        value: Number(head) * minority,
        currency,
        minority,
        negativeSymbol: 'hyphen-minus',
    });

    // Так как -0 === 0, formatAmount возвращает положительное значение. Исправляем это здесь
    if (/^-(,|0),?/.test(enteredValue) && majorPart === '0') {
        majorPart = `-${majorPart}`;
    }

    if (!tail && enteredValue.includes(',')) {
        return majorPart.concat(',');
    }

    if (tail) {
        return majorPart.concat(',', tail.slice(0, minority.toString().length - 1));
    }

    return majorPart;
}

export function getAmountValueFromStr(str: string, minority: number) {
    if (str === '' || str === '-') {
        return null;
    }

    return Math.round(Number(str.replace(',', '.').replace(/[^0-9.-]/g, '')) * minority);
}
