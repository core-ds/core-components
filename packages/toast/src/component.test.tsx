import React, { forwardRef, ForwardRefRenderFunction } from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import * as popoverModule from '@alfalab/core-components-popover';
import { act } from 'react-dom/test-utils';
import { ToastDesktop as Toast, ToastDesktopProps as ToastProps } from './desktop';

import { asyncRender } from '../../utils/test-utils';

type PopoverComponent = {
    render?: ForwardRefRenderFunction<HTMLDivElement, popoverModule.PopoverProps>;
};

describe('Toast', () => {
    jest.useFakeTimers();

    const baseProps = {
        open: true,
        onClose: jest.fn(),
    };

    const getPortalContainer = () => {
        if (document.querySelector('#portalContainer')) {
            return document.querySelector('#portalContainer') as HTMLElement;
        }

        const portalContainer = document.createElement('div');
        portalContainer.setAttribute('id', 'portalContainer');
        document.body.appendChild(portalContainer);
        return portalContainer;
    };

    describe('Snapshots tests', () => {
        it('should match snapshot when open', () => {
            const { baseElement } = render(<Toast {...baseProps}>text</Toast>);

            expect(baseElement).toMatchSnapshot();
        });

        it('should match snapshot when closed', () => {
            const { baseElement } = render(
                <Toast {...baseProps} open={false}>
                    text
                </Toast>,
            );

            expect(baseElement).toMatchSnapshot();
        });

        it('should math snapshot when prop `anchorElement` is passed', async () => {
            const anchorElement = document.createElement('div');
            document.body.appendChild(anchorElement);

            const { baseElement } = await asyncRender(
                <Toast {...baseProps} anchorElement={anchorElement}>
                    text
                </Toast>,
            );

            expect(baseElement).toMatchSnapshot();
        });

        it('should math snapshot when prop `anchorElement` and `useAnchorWidth` is passed', async () => {
            const anchorElement = document.createElement('div');
            anchorElement.style.width = '100px';

            document.body.appendChild(anchorElement);

            const { baseElement } = await asyncRender(
                <Toast {...baseProps} anchorElement={anchorElement} useAnchorWidth={true}>
                    text
                </Toast>,
            );

            expect(baseElement).toMatchSnapshot();
        });
    });

    it('should set `className` class', () => {
        const className = 'test-class';

        render(
            <Toast {...baseProps} className={className} getPortalContainer={getPortalContainer} />,
        );

        expect(getPortalContainer().firstElementChild).toHaveClass(className);
    });

    it('should pass props to Popover', async () => {
        const PopoverComponent = popoverModule.Popover as PopoverComponent;

        const popoverComponentSpy = jest.spyOn(PopoverComponent, 'render');

        const anchorElement = document.createElement('div');
        document.body.appendChild(anchorElement);

        const popoverProps: Partial<ToastProps> = {
            position: 'top',
            offset: [5, 5],
            open: true,
            getPortalContainer,
            anchorElement,
            /*
             * todo: почему-то preventFlip и transition не прокидываются, хотя указаны в типах пропсов. Нужно либо прокинуть либо убрать из типов
             * preventFlip: false,
             * transition: { timeout: 200 }
             */
        };

        await act(async () => {
            await render(<Toast {...baseProps} anchorElement={anchorElement} {...popoverProps} />);
        });

        const popoverLastCall =
            popoverComponentSpy.mock.calls[popoverComponentSpy.mock.calls.length - 1];

        expect(popoverLastCall[0]).toMatchObject(popoverProps);
    });

    it('should set bottomOffset', () => {
        render(<Toast {...baseProps} bottomOffset={10} getPortalContainer={getPortalContainer} />);

        const toastPlate = getPortalContainer().firstElementChild as HTMLElement;

        expect(toastPlate).toHaveStyle('bottom: 10px');
    });

    it('should forward ref to ToastPlate wrapper', () => {
        const ref = jest.fn();

        render(<Toast {...baseProps} ref={ref} />);

        expect(ref.mock.calls[0][0].tagName).toBe('DIV');
    });

    describe('Callback tests', () => {
        it('should call onClose when click outside', async () => {
            const onClose = jest.fn();
            render(<Toast onClose={onClose} open={true} getPortalContainer={getPortalContainer} />);

            fireEvent.click(getPortalContainer().firstElementChild as HTMLElement);
            expect(onClose).not.toBeCalled();

            await userEvent.setup({ delay: null }).click(document.body);
            expect(onClose).toBeCalled();
        });

        it('should not call onClose when click outside', async () => {
            const onClose = jest.fn();
            render(
                <Toast
                    onClose={onClose}
                    open={true}
                    getPortalContainer={getPortalContainer}
                    closeWithClickOutside={false}
                />,
            );

            fireEvent.click(getPortalContainer().firstElementChild as HTMLElement);
            expect(onClose).not.toBeCalled();

            await userEvent.setup({ delay: null }).click(document.body);
            expect(onClose).not.toBeCalled();
        });

        it('should call onClose after delay', () => {
            const onClose = jest.fn();
            render(<Toast onClose={onClose} open={true} autoCloseDelay={3000} />);

            jest.advanceTimersByTime(2500);
            expect(onClose).not.toBeCalled();

            jest.advanceTimersByTime(3500);
            expect(onClose).toBeCalled();
        });

        it('should not call onClose if mouse is over ToastPlate', async () => {
            const onClose = jest.fn();
            render(
                <Toast
                    onClose={onClose}
                    open={true}
                    autoCloseDelay={3000}
                    getPortalContainer={getPortalContainer}
                />,
            );

            const toastPlate = getPortalContainer().firstElementChild as HTMLElement;

            jest.advanceTimersByTime(2500);
            await userEvent.setup({ delay: null }).hover(toastPlate);
            jest.advanceTimersByTime(5000);

            expect(onClose).not.toBeCalled();
        });

        it('should call onClose after mouse left ToastPlate', () => {
            const onClose = jest.fn();
            render(
                <Toast
                    onClose={onClose}
                    open={true}
                    autoCloseDelay={3000}
                    getPortalContainer={getPortalContainer}
                />,
            );

            const toastPlate = getPortalContainer().firstElementChild as HTMLElement;

            jest.advanceTimersByTime(2500);
            userEvent.hover(toastPlate);
            userEvent.unhover(toastPlate);
            jest.advanceTimersByTime(5000);

            expect(onClose).toBeCalled();
        });

        it('should not call onClose if touch ToastPlate', () => {
            const onClose = jest.fn();
            render(
                <Toast
                    onClose={onClose}
                    open={true}
                    autoCloseDelay={3000}
                    getPortalContainer={getPortalContainer}
                />,
            );

            const toastPlate = getPortalContainer().firstElementChild as HTMLElement;

            jest.advanceTimersByTime(2500);
            fireEvent.touchStart(toastPlate);
            jest.advanceTimersByTime(5000);

            expect(onClose).not.toBeCalled();
        });

        it('should render custom toast plate', async () => {
            const dataTestId = 'testId';
            const onClose = jest.fn();

            const CustomToastPlate: ToastProps['ToastPlate'] = forwardRef((_, ref: any) => (
                <div ref={ref} data-test-id={dataTestId} />
            ));

            const { getByTestId } = render(
                <Toast
                    onClose={onClose}
                    open={true}
                    autoCloseDelay={3000}
                    getPortalContainer={getPortalContainer}
                    ToastPlate={CustomToastPlate}
                />,
            );

            await waitFor(() => expect(getByTestId(dataTestId)).toBeInTheDocument());
        });
    });

    it('should unmount without errors', () => {
        const { unmount } = render(<Toast {...baseProps} />);

        expect(unmount).not.toThrowError();
    });
});
