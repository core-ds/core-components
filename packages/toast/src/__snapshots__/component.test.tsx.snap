// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast Snapshots tests should match snapshot when closed 1`] = `
<body>
  <div
    alfa-portal-container=""
  />
  <div />
</body>
`;

exports[`Toast Snapshots tests should match snapshot when open 1`] = `
<body>
  <div />
  <div
    alfa-portal-container=""
  >
    <div
      aria-atomic="true"
      aria-live="assertive"
      class="component component fixed toastPlate"
      role="alert"
      style="z-index: 1000;"
    >
      <div
        class="wrapper"
      >
        <div
          class="contentWrapper"
        >
          <div
            class="content"
          >
            <div
              class="contentTitle"
            >
              <div
                class="children"
              >
                text
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Toast Snapshots tests should math snapshot when prop \`anchorElement\` and \`useAnchorWidth\` is passed 1`] = `
<body>
  <div
    alfa-portal-container=""
  >
    <div
      class="component"
      data-popper-placement="left"
      style="z-index: 1000; position: fixed; left: 0px; top: 0px; transform: translate(0px, 0px); width: 0px;"
    >
      <div
        class="inner popoverInner"
      >
        <div
          class=""
        >
          <div
            aria-atomic="true"
            aria-live="assertive"
            class="component component"
            role="alert"
          >
            <div
              class="wrapper"
            >
              <div
                class="contentWrapper"
              >
                <div
                  class="content"
                >
                  <div
                    class="contentTitle"
                  >
                    <div
                      class="children"
                    >
                      text
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div />
  <div
    style="width: 100px;"
  />
  <div />
</body>
`;

exports[`Toast Snapshots tests should math snapshot when prop \`anchorElement\` is passed 1`] = `
<body>
  <div
    alfa-portal-container=""
  >
    <div
      class="component"
      data-popper-placement="left"
      style="z-index: 1000; position: fixed; left: 0px; top: 0px; transform: translate(0px, 0px);"
    >
      <div
        class="inner popoverInner"
      >
        <div
          class=""
        >
          <div
            aria-atomic="true"
            aria-live="assertive"
            class="component component"
            role="alert"
          >
            <div
              class="wrapper"
            >
              <div
                class="contentWrapper"
              >
                <div
                  class="content"
                >
                  <div
                    class="contentTitle"
                  >
                    <div
                      class="children"
                    >
                      text
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div />
  <div />
</body>
`;
